In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import zipfile
import os

# Шлях до zip-файлу
zip_path = "C:\\Users\\Igor\\Downloads\\homework.zip"#Сюди вставляєте ваш шлях саме до zip файлу
extract_folder = '/new/homework'#Тут можете залишити усе без змі

# Розпакування файл
with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    zip_ref.extractall(extract_folder)

# Список файлів у розпакованій папці
extracted_files = os.listdir(extract_folder)
extracted_files[0]

'data'

In [13]:
# Перерахування файлів в папці "data".
data_folder_path = os.path.join(extract_folder, extracted_files[0])
data_folder_path

'/new/homework\\data'

In [14]:

data_files = os.listdir(data_folder_path)
data_files

['idle', 'running', 'stairs', 'walking']

In [18]:
# Функція для завантаження даних із папки та функцій обчислення
def load_and_compute_features(folder_path):
    files = os.listdir(folder_path)
    features_list = []

    for file in files:
        file_path = os.path.join(folder_path, file)
        data = pd.read_csv(file_path)

        # Функції обчислювальної часової області
        features = {
            'mean_X': data['accelerometer_X'].mean(),
            'mean_Y': data['accelerometer_Y'].mean(),
            'mean_Z': data['accelerometer_Z'].mean(),
            'std_X': data['accelerometer_X'].std(),
            'std_Y': data['accelerometer_Y'].std(),
            'std_Z': data['accelerometer_Z'].std(),
            'min_X': data['accelerometer_X'].min(),
            'min_Y': data['accelerometer_Y'].min(),
            'min_Z': data['accelerometer_Z'].min(),
            'max_X': data['accelerometer_X'].max(),
            'max_Y': data['accelerometer_Y'].max(),
            'max_Z': data['accelerometer_Z'].max(),
            'range_X': data['accelerometer_X'].max() - data['accelerometer_X'].min(),
            'range_Y': data['accelerometer_Y'].max() - data['accelerometer_Y'].min(),
            'range_Z': data['accelerometer_Z'].max() - data['accelerometer_Z'].min(),
            'median_X': data['accelerometer_X'].median(),
            'median_Y': data['accelerometer_Y'].median(),
            'median_Z': data['accelerometer_Z'].median(),
            'amplitude_X': data['accelerometer_X'].max() - data['accelerometer_X'].min(),
            'amplitude_Y': data['accelerometer_Y'].max() - data['accelerometer_Y'].min(),
            'amplitude_Z': data['accelerometer_Z'].max() - data['accelerometer_Z'].min(),
        }
        features_list.append(features)
        print(f"len(features_list) = {len(features_list)}")
    return pd.DataFrame(features_list)

# Функції обчислення для кожного типу діяльності
features_idle = load_and_compute_features(os.path.join(data_folder_path, data_files[0]))
features_running = load_and_compute_features(os.path.join(data_folder_path, data_files[1]))
features_stairs = load_and_compute_features(os.path.join(data_folder_path, data_files[2]))
features_walking = load_and_compute_features(os.path.join(data_folder_path, data_files[3]))

# Об’єднання всіх функції в єдиний DataFrame з мітками
features_idle['activity'] = data_files[0]
features_running['activity'] = data_files[1]
features_stairs['activity'] = data_files[2]
features_walking['activity'] = data_files[3]

combined_features = pd.concat([features_idle, features_running, features_stairs, features_walking], ignore_index=True)
#combined_features
features_idle

len(features_list) = 1
len(features_list) = 2
len(features_list) = 3
len(features_list) = 4
len(features_list) = 5
len(features_list) = 6
len(features_list) = 7
len(features_list) = 8
len(features_list) = 9
len(features_list) = 10
len(features_list) = 11
len(features_list) = 12
len(features_list) = 13
len(features_list) = 14
len(features_list) = 15
len(features_list) = 16
len(features_list) = 17
len(features_list) = 18
len(features_list) = 19
len(features_list) = 20
len(features_list) = 21
len(features_list) = 22
len(features_list) = 23
len(features_list) = 24
len(features_list) = 25
len(features_list) = 26
len(features_list) = 27
len(features_list) = 28
len(features_list) = 29
len(features_list) = 30
len(features_list) = 31
len(features_list) = 32
len(features_list) = 33
len(features_list) = 34
len(features_list) = 35
len(features_list) = 36
len(features_list) = 37
len(features_list) = 38
len(features_list) = 39
len(features_list) = 40
len(features_list) = 41
len(features_list) = 42
l

mean_X    mean_Y    mean_Z     std_X     std_Y     std_Z     min_X  \
0     0.178448  0.167435  9.605697  1.036361  1.157603  0.406903 -0.909797   
1    -0.098641 -0.131202  9.771216  0.125848  0.048059  0.025419 -0.320823   
2    -0.099918  0.220905  9.768503  0.015642  0.016383  0.014528 -0.124498   
3     0.400949 -0.029528  9.752541  0.052889  0.054525  0.017967  0.268151   
4     0.400949 -0.027134  9.753659  0.053187  0.054004  0.018902  0.268151   
...        ...       ...       ...       ...       ...       ...       ...   
1034  0.407972 -0.028411  9.755255  0.034722  0.037089  0.015953  0.330400   
1035  0.407812 -0.026815  9.753339  0.034509  0.038030  0.016765  0.330400   
1036  0.401747 -0.026336  9.753020  0.038294  0.038467  0.016074  0.320823   
1037  0.403822 -0.026017  9.753180  0.037374  0.038976  0.016108  0.320823   
1038  0.400630 -0.027134  9.752062  0.036453  0.037899  0.015936  0.320823   

         min_Y     min_Z     max_X  ...   range_X   range_Y   range_Z  \
0    -0.282516  8.418014  5.099650  ...  6.009447  4.898537  1.388636   
1    -0.244209  9.667787  0.407014  ...  0.727837  0.268151  0.138863   
2     0.177171  9.739613 -0.062249  ...  0.062249  0.067038  0.052672   
3    -0.205901  9.725247  0.560243  ...  0.292092  0.292092  0.071826   
4    -0.205901  9.725247  0.560243  ...  0.292092  0.292092  0.071826   
...        ...       ...       ...  ...       ...       ...       ...   
1034 -0.105345  9.730036  0.488417  ...  0.158017  0.162806  0.057461   
1035 -0.105345  9.725247  0.488417  ...  0.158017  0.162806  0.062250   
1036 -0.105345  9.725247  0.488417  ...  0.167594  0.162806  0.062250   
1037 -0.105345  9.725247  0.488417  ...  0.167594  0.162806  0.062250   
1038 -0.105345  9.725247  0.488417  ...  0.167594  0.162806  0.062250   

      median_X  median_Y  median_Z  amplitude_X  amplitude_Y  amplitude_Z  \
0    -0.102950 -0.131681  9.770737     6.009447     4.898537     1.388636   
1    -0.114922 -0.131681  9.777920     0.727837     0.268151     0.138863   
2    -0.100556  0.225055  9.768343     0.062249     0.067038     0.052672   
3     0.392649 -0.021548  9.751583     0.292092     0.292092     0.071826   
4     0.402226 -0.019154  9.751583     0.292092     0.292092     0.071826   
...        ...       ...       ...          ...          ...          ...   
1034  0.404620 -0.028730  9.753978     0.158017     0.162806     0.057461   
1035  0.404620 -0.023942  9.749189     0.158017     0.162806     0.062250   
1036  0.395043 -0.019154  9.751583     0.167594     0.162806     0.062250   
1037  0.404620 -0.023942  9.751583     0.167594     0.162806     0.062250   
1038  0.395043 -0.021548  9.751583     0.167594     0.162806     0.062250   

      activity  
0         idle  
1         idle  
2         idle  
3         idle  
4         idle  
...        ...  
1034      idle  
1035      idle  
1036      idle  
1037      idle  
1038      idle  

[1039 rows x 22 columns]

In [19]:
features_stairs

mean_X     mean_Y    mean_Z     std_X     std_Y     std_Z     min_X  \
0    2.830745  -7.161376 -0.104068  3.302172  4.710145  4.512922 -4.582502   
1    2.033795  -9.477846 -0.449471  3.476321  4.791658  5.165997 -4.649539   
2    0.864945  -9.368990 -3.801513  2.847727  4.230074  6.047028 -6.224924   
3    0.850101  -9.412564 -3.967192  2.812559  4.164104  5.935387 -6.224924   
4    0.570139  -8.606676 -3.642538  2.715524  3.422683  5.846725 -6.224924   
..        ...        ...       ...       ...       ...       ...       ...   
160  0.881545  -9.970413 -2.812389  2.962215  4.482366  6.292407 -6.224924   
161  0.871490  -9.944875 -3.237120  2.928768  4.410401  6.027437 -6.224924   
162  0.881705 -10.013190 -3.722505  2.938536  4.417071  6.223244 -6.224924   
163  0.964544 -10.231062 -3.843811  2.958615  4.324824  6.268747 -6.224924   
164  0.957681  -9.805852 -3.457866  2.932814  4.529009  6.319447 -6.224924   

         min_Y      min_Z     max_X  ...    range_X    range_Y    range_Z  \
0   -17.300500 -14.422671  8.073248  ...  12.655750  20.274099  25.005043   
1   -19.416977 -14.422671  8.288726  ...  12.938265  18.933348  25.005043   
2   -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
3   -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
4   -17.501614 -26.230873  8.398860  ...  14.623784  14.786589  33.006464   
..         ...        ...       ...  ...        ...        ...        ...   
160 -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
161 -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
162 -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
163 -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   
164 -22.720974 -26.230873  8.398860  ...  14.623784  20.005949  33.006464   

     median_X   median_Y  median_Z  amplitude_X  amplitude_Y  amplitude_Z  \
0    3.447650  -7.153874  0.019154    12.655750    20.274099    25.005043   
1    1.632846  -7.963115 -0.378283    12.938265    18.933348    25.005043   
2    1.376666  -9.289502 -3.330335    14.623784    20.005949    33.006464   
3    1.132457  -9.265560 -3.330335    14.623784    20.005949    33.006464   
4    0.773327  -8.482656 -3.186682    14.623784    14.786589    33.006464   
..        ...        ...       ...          ...          ...          ...   
160  1.235408 -10.110713 -2.454057    14.623784    20.005949    33.006464   
161  1.132457  -9.837775 -3.078944    14.623784    20.005949    33.006464   
162  1.132457  -9.837775 -3.584120    14.623784    20.005949    33.006464   
163  1.376666 -10.110713 -3.584120    14.623784    20.005949    33.006464   
164  1.376666  -9.536105 -3.078944    14.623784    20.005949    33.006464   

     activity  
0      stairs  
1      stairs  
2      stairs  
3      stairs  
4      stairs  
..        ...  
160    stairs  
161    stairs  
162    stairs  
163    stairs  
164    stairs  

[165 rows x 22 columns]

In [21]:
# Вибір ознак і цільової змінної
X = combined_features.drop("activity", axis=1)  # Ознаки
y = combined_features["activity"]  # Цільова змінна
X

mean_X     mean_Y    mean_Z     std_X     std_Y     std_Z     min_X  \
0     0.178448   0.167435  9.605697  1.036361  1.157603  0.406903 -0.909797   
1    -0.098641  -0.131202  9.771216  0.125848  0.048059  0.025419 -0.320823   
2    -0.099918   0.220905  9.768503  0.015642  0.016383  0.014528 -0.124498   
3     0.400949  -0.029528  9.752541  0.052889  0.054525  0.017967  0.268151   
4     0.400949  -0.027134  9.753659  0.053187  0.054004  0.018902  0.268151   
...        ...        ...       ...       ...       ...       ...       ...   
6457  0.396639 -10.702081 -0.202230  3.474121  5.285310  8.666662 -5.990292   
6458  0.418028 -10.736398 -0.090980  3.533930  5.329992  8.675810 -7.359776   
6459  1.151771 -10.878295  0.364557  3.948034  5.412533  7.863650 -7.359776   
6460  1.551123 -11.146126  0.233036  4.016935  5.294201  7.943439 -7.359776   
6461  1.131979 -10.937192  0.260170  4.212427  5.443090  7.941497 -7.359776   

          min_Y      min_Z      max_X  ...      max_Z    range_X    range_Y  \
0     -0.282516   8.418014   5.099650  ...   9.806650   6.009447   4.898537   
1     -0.244209   9.667787   0.407014  ...   9.806650   0.727837   0.268151   
2      0.177171   9.739613  -0.062249  ...   9.792285   0.062249   0.067038   
3     -0.205901   9.725247   0.560243  ...   9.797073   0.292092   0.292092   
4     -0.205901   9.725247   0.560243  ...   9.797073   0.292092   0.292092   
...         ...        ...        ...  ...        ...        ...        ...   
6457 -25.201366 -20.058622   7.192182  ...  32.599450  13.182474  25.383325   
6458 -25.201366 -20.058622   7.192182  ...  32.599450  14.551958  25.383325   
6459 -25.201366 -16.903065  11.113884  ...  32.599450  18.473660  25.383325   
6460 -25.201366 -16.903065  11.113884  ...  32.599450  18.473660  25.383325   
6461 -25.201366 -16.903065  11.113884  ...  32.599450  18.473660  25.383325   

        range_Z  median_X   median_Y  median_Z  amplitude_X  amplitude_Y  \
0      1.388636 -0.102950  -0.131681  9.770737     6.009447     4.898537   
1      0.138863 -0.114922  -0.131681  9.777920     0.727837     0.268151   
2      0.052672 -0.100556   0.225055  9.768343     0.062249     0.067038   
3      0.071826  0.392649  -0.021548  9.751583     0.292092     0.292092   
4      0.071826  0.402226  -0.019154  9.751583     0.292092     0.292092   
...         ...       ...        ...       ...          ...          ...   
6457  52.658072  0.390255 -10.414777  0.833182    13.182474    25.383325   
6458  52.658072  0.390255 -10.414777  0.847548    14.551958    25.383325   
6459  49.502515  1.501165 -10.414777  0.833182    18.473660    25.383325   
6460  49.502515  1.889025 -10.747571  0.768539    18.473660    25.383325   
6461  49.502515  1.501165 -10.414777  0.768539    18.473660    25.383325   

      amplitude_Z  
0        1.388636  
1        0.138863  
2        0.052672  
3        0.071826  
4        0.071826  
...           ...  
6457    52.658072  
6458    52.658072  
6459    49.502515  
6460    49.502515  
6461    49.502515  

[6462 rows x 21 columns]

In [22]:
y

0          idle
1          idle
2          idle
3          idle
4          idle
         ...   
6457    walking
6458    walking
6459    walking
6460    walking
6461    walking
Name: activity, Length: 6462, dtype: object

In [23]:
# Розділення на тренувальний та тестовий набори
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)
X_train

mean_X     mean_Y    mean_Z     std_X      std_Y     std_Z      min_X  \
1413  6.792829   7.066088  2.840800  9.549495   8.335001  7.194899  -3.045424   
2201  4.961744  10.945492  2.696350  7.789352  13.974918  6.901295  -6.133945   
3753  6.123091   9.185594  5.152003  7.519555  10.204258  8.287980  -5.631162   
4812 -4.240131  -9.518867 -1.281536  4.002251   4.264463  8.531402 -13.421895   
869  -0.375570   1.811613  8.987674  2.022436   2.572131  1.223811 -10.448297   
...        ...        ...       ...       ...        ...       ...        ...   
3772  5.619990   7.580043  1.711854  7.871821  13.107397  5.317100  -6.315904   
5191 -2.201708  -9.175858 -1.480574  2.341183   5.122290  2.746525  -8.470686   
5226  0.338540 -11.188743 -1.286484  5.299653   8.208943  7.546326 -11.798626   
5390 -1.742340 -10.166738 -1.533406  3.882006   6.662437  9.135771  -9.504981   
860   0.270066  -0.002554  9.772493  0.137047   0.089739  0.029963   0.023942   

          min_Y      min_Z      max_X  ...      max_Z    range_X    range_Y  \
1413 -10.261548  -6.033388  32.953793  ...  19.292477  35.999217  33.782186   
2201 -12.622231 -11.291056  27.758375  ...  21.145590  33.892320  51.580681   
3753 -14.302961  -7.857770  23.573310  ...  25.565287  29.204472  41.424478   
4812 -16.524780 -17.937359   3.323152  ...  32.312145  16.745047  16.357186   
869   -1.996764   4.553772   0.541090  ...   9.797073  10.989387   7.991845   
...         ...        ...        ...  ...        ...        ...        ...   
3772 -16.342821  -7.292738  23.755270  ...  14.427460  30.071174  50.091491   
5191 -20.556616  -8.212111   2.226608  ...   2.911349  10.697294  21.835120   
5226 -27.983427 -25.292345   9.327810  ...  12.401965  21.126436  39.365462   
5390 -23.376986 -26.302700   7.527370  ...  21.648373  17.032351  26.800695   
860   -0.201113   9.701305   0.938527  ...   9.873688   0.914585   0.612916   

        range_Z  median_X   median_Y  median_Z  amplitude_X  amplitude_Y  \
1413  25.325865  3.636792   7.797915 -0.229844    35.999217    33.782186   
2201  32.436646  3.892972  10.381259  1.244985    33.892320    51.580681   
3753  33.423057  5.449204   8.188170  3.394978    29.204472    41.424478   
4812  50.249504 -2.880224  -9.521740 -0.823605    16.745047    16.357186   
869    5.243301  0.220267   0.014365  9.742007    10.989387     7.991845   
...         ...       ...        ...       ...          ...          ...   
3772  21.720198  4.247314   7.548918  0.761356    30.071174    50.091491   
5191  11.123460 -2.066196  -9.485827 -0.644040    10.697294    21.835120   
5226  37.694310 -0.464475 -12.212823  0.124499    21.126436    39.365462   
5390  47.951073 -2.001552  -9.531318 -1.113304    17.032351    26.800695   
860    0.172383  0.256179  -0.009577  9.773131     0.914585     0.612916   

      amplitude_Z  
1413    25.325865  
2201    32.436646  
3753    33.423057  
4812    50.249504  
869      5.243301  
...           ...  
3772    21.720198  
5191    11.123460  
5226    37.694310  
5390    47.951073  
860      0.172383  

[3877 rows x 21 columns]

In [24]:
import time
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

# Навчання моделі SVM
start_time = time.time()
svm_model = SVC(kernel='linear').fit(X_train, y_train)
svm_training_time = time.time() - start_time
svm_model


SVC(kernel='linear')

In [26]:
# Навчання моделі випадкового лісу
start_time = time.time()
rf_model = RandomForestClassifier(n_estimators=100, random_state=42).fit(X_train, y_train)
rf_training_time = time.time() - start_time
rf_model


RandomForestClassifier(random_state=42)

In [27]:
# Оцінка результатів
svm_predictions = svm_model.predict(X_test)
rf_predictions = rf_model.predict(X_test)

# Отримання звіту про класифікацію
svm_report = classification_report(y_test, svm_predictions)
rf_report = classification_report(y_test, rf_predictions)

print("SVM classification report:")
print("Training time:", svm_training_time, "seconds")
print(svm_report)

print("\nRandom Forest classification report:")
print("Training time:", rf_training_time, "seconds")
print(rf_report)

SVM classification report:
Training time: 0.07399845123291016 seconds
              precision    recall  f1-score   support

        idle       1.00      1.00      1.00       442
     running       1.00      1.00      1.00      1360
      stairs       0.86      0.75      0.80        64
     walking       0.98      0.99      0.98       719

    accuracy                           0.99      2585
   macro avg       0.96      0.93      0.95      2585
weighted avg       0.99      0.99      0.99      2585


Random Forest classification report:
Training time: 0.5410082340240479 seconds
              precision    recall  f1-score   support

        idle       1.00      1.00      1.00       442
     running       1.00      1.00      1.00      1360
      stairs       1.00      0.89      0.94        64
     walking       0.99      1.00      1.00       719

    accuracy                           1.00      2585
   macro avg       1.00      0.97      0.98      2585
weighted avg       1.00      1.00  

Зі звіту SVM classification report можна отримати наступні висновки:

Accuracy (Точність): Загальна точність класифікації складає 99%, що означає, що модель правильно класифікувала 99% прикладів із тестового набору даних.
Precision (Точність): Точність визначає, яка частка елементів, визначених класифікатором як певний клас, насправді належить до цього класу. З цього звіту видно, що точність для кожного класу є дуже високою (більше 0.86), що означає, що класифікатор досить точно визначає кожний клас.
Recall (Повнота): Повнота показує, яка частка елементів певного класу була визначена правильно класифікатором. Також високий рівень повноти (більше 0.75) спостерігається для кожного класу, що означає, що класифікатор ефективно виявляє екземпляри кожного класу.
F1-score: F1-score є середнім гармонічним між точністю і повнотою. Це особливо корисна метрика для незбалансованих класів. Значення F1-score близько до 1 для кожного класу, що свідчить про те, що модель має хороший баланс між точністю і повнотою для всіх класів.
Support (Підтримка): Кількість прикладів у кожному класі.
Macro avg і weighted avg: Макро-середнє і зважене середнє використовується для обчислення середньозважених значень для метрик. Macro avg обчислюється, обчислюючи середнє значення для кожного класу без врахування його розміру, тоді як weighted avg враховує розмір кожного класу при обчисленні середньозважених з

За наведеним звітом Random Forest classification report можна зробити такі висновки:

Точність моделі: Загальна точність класифікації становить 100%, що означає, що модель правильно класифікувала всі приклади з тестового набору даних.
Повнота (Recall): Повнота для кожного класу також дуже висока, з винятком класу "stairs", де вона дорівнює 0.89. Це означає, що модель ефективно виявляє більшість прикладів для кожного класу, але може бути деяка недосконалість у виявленні прикладів класу "stairs".
F1-score: F1-score, який є середнім гармонічним між точністю і повнотою, також є високим для кожного класу, з винятком класу "stairs", де він становить 0.94. Це підтверджує ефективність моделі у класифікації, але також вказує на можливість поліпшення у виявленні прикладів класу "stairs".
Підтримка (Support): Кількість прикладів у кожному класі показує, що даний звіт базується на значної кількості даних для кожного класу, що дозволяє робити висновки про ефективність моделі для кожного класу.
Macro avg і weighted avg: Обидва середні значення підтверджують високу ефективність моделі для всього набору даних, демонструючи високу точність і повноту для всіх класів.
Отже, загальний висновок полягає в тому, що модель Random Forest є дуже ефективною для класифікації даних, з високою точністю, повнотою і F1-score для кожного класу. Однак можливо, що модель може бути трохи менш ефективною у виявленні прикладів класу "stairs", що може вимагати додаткового удоскон

За результатами звітів з класифікації SVM та Random Forest можна зробити наступні висновки:

Точність (Accuracy):
SVM: Точність класифікації для SVM становить 99%.
Random Forest: Точність класифікації для Random Forest становить 100%.
Повнота (Recall):
SVM: Повнота для SVM варіюється від 0.75 до 1.00 для різних класів.
Random Forest: Повнота для Random  також варіюється від 0.89 до 1.00 для різних класів.
F1-score:
SVM: F1-score для SVM також варіюється від 0.80 до 1.00 для різних класів.
Random Forest: F1-score для Random Forest також варіюється від 0.94 до 1.00 для різних класів.
Середнє значення:
SVM: Макро avg та weighted avg для SVM показують значення 0.96 та 0.99 відповідно.
Random Forest: Макро avg та weighted avg для Random Forest показують значення 1.00 для обох.
Загалом, обидва класифікатори демонструють високу ефективність у класифікації даних, але Random Forest виявляється трохи ефективнішим за SVM за кількістю точних класифікацій та середніми значеннями метрик.алення.начень.